text/0000-structural-records.md

```diff
@@ -0,0 +1,1298 @@
+  They are called product types because the number of possible values that the
+  type can take is the *product* of the number of values that each component /
+  factor / operand type can take. For example, if we consider `(A, B, C)`,
+  the number of values is `values(A) * values(B) * values(C)`.
+
+- A *[record type]* is a special case of a product type in which each component
+  type is *labeled*. In Rust, record types are `struct`s with named fields.
+  For example, you can write:
+
+  ```rust
+  struct Foo {
+      bar: u8,
+      baz: bool,
+  }
+  ```
+
+  The only case of a record type in Rust uses nominal typing.
+  There is currently no structural variant of record types.
+  This brings us to the proposal in this RFC...
+
+## Proposal
+
+As we've seen, Rust currently lacks a structurally typed variant of record types.
+In this RFC, we propose to change this by introducing record types that are
+also structural, or in other words: *"structural records"*.
+
+### Construction
+
+To create a structural record with the field `alpha` with value `42u8`,
+`beta` with value `true`, and `gamma` with value `"Dancing Ferris"`,
+you can simply write:
+
+```rust
+let my_record = {
+    alpha: 42u8,
+    beta: true,
+    gamma: "Dancing Ferris"
+};
+```
+
+Note how this is the same syntax as used for creating normal structs but without
+the name of the struct. So we have taken:
+
+```rust
+struct MyRecordType {
+    alpha: u8,
+    beta: bool,
+    gamma: &'static str,
+}
+
+let my_record_nominal = MyRecordType {
+    alpha: 42u8,
+    beta: true,
+    gamma: "Dancing Ferris"
+};
+```
+
+and removed `MyRecordType`. Note that because we are using structural typing,
+we did not have to define a type `MyRecordType` ahead of time.
+
+If you already had variables named `alpha` and `beta` in scope,
+just as you could have with `MyRecordType`, you could have also written:
+
+```rust
+let alpha = 42u8;
+let beta = true;
+
+let my_record = {
+    alpha,
+    beta,
+    gamma: "Dancing Ferris"
+};
+```
+
+### Pattern matching
+
+Once you have produced a structural record, you can also pattern match on the
+expression. To do so, you can write:
+
+```rust
+match my_record {
+    { alpha, beta, gamma } => println!("{}, {}, {}", alpha, beta, gamma),
+}
+```
+
+This pattern is *irrefutable* so you can also just write:
+
+```rust
+let { alpha, beta, gamma } = my_record;
+println!("{}, {}, {}", alpha, beta, gamma);
+```
+
+This is not particular to `match` and `let`. This also works for `if let`,
+`while let`, `for` loops, and function arguments.
+
+If we had used `MyRecordType`, you would have instead written:
+
+```rust
+let MyRecordType { alpha, beta, gamma } = my_record_nominal;
+println!("{}, {}, {}", alpha, beta, gamma);
+```
+
+When pattern matching on a structural record, it is also possible to give
+the binding you've created a different name. To do so, write:
+
+```rust
+let { alpha: new_alpha, beta, gamma } = my_record;
```

Comment:
I think this is easier to understand if the different name is on the left. Example:

```rust
let { new_alpha: alpha, beta, gamma } = my_record;
```
Also, using the `:` makes the right side look like a type.  Would using `=` work better? Example:

```rust
let { new_alpha = alpha, beta, gamma } = my_record;
```

Reply:
>Also, using the : makes the right side look like a type. Would using = work better? 

Of course it would, but rust did it wrong and now it would be better to share conformity with the struct instance colon syntax.